EWS should use results database instead of doing clean-tree runs

aj062 opened the pull request:
#### b9824fc290ac9247e8a04f38632a8fdd4b9ec02d
<pre>
EWS should use results database instead of doing clean-tree runs
<a href="https://bugs.webkit.org/show_bug.cgi?id=204368">https://bugs.webkit.org/show_bug.cgi?id=204368</a>
rdar://82975679

Reviewed by Jonathan Bedard.

Pre-existing test failures signficantly slow down EWS, and this change attempts to speed up EWS
by consulting results database. EWS will now consult results database for the test failures which were noticed.
If those test were also failing/flaky in results db, then we ignore those failures in EWS run. Othewise, EWS will
continue with the existing logic of retry/clean-tree-run.

* Tools/CISupport/ews-build/results_db.py:
(ResultsDatabase.get_results_summary): Updated the method to not print logs, instead return them to the caller.
(ResultsDatabase.is_test_flaky_or_pre_existing): Wrapper method which contains the logic to declare if test is flaky/pre-existing or not.
* Tools/CISupport/ews-build/steps.py:
(RunWebKitTests.commandComplete):
(RunWebKitTests.filter_failures_using_results_db): Method which filters the test failures based on info from results db.
(RunWebKitTests.evaluateCommand):
(RunWebKitTests.getResultSummary):
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/257090@main">https://commits.webkit.org/257090@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f6f78d980b85cce4aac8c9053ec7975e1605aadd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97878 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/7105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/31050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/107355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/167632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101819 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/7554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/35879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/103992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103519 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/7554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/84496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/90249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/7554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/31050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/1088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/31050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/96806 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/1075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/31050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/5904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/84496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2421 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2346 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/31050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/CISupport/ews-build/results_db.py:
```diff
@@ -48,25 +50,46 @@ class ResultsDatabase(object):
+        is_existing_failure = (pass_rate <= cls.PERECENT_SUCCESS_RATE_FOR_PRE_EXISTING_FAILURE)
+        is_flaky_failure = (pass_rate <= cls.PERECENT_SUCCESS_RATE_FOR_FLAKY_FAILURE) and not is_existing_failure
+        output = {
+            'is_existing_failure': is_existing_failure,
```
I'm a bit unclear what benefit we get by trying to distinguish `flakey` and `existing` failures, because to EWS, they look very similar.

Consider 2 examples:
```
                    V
p-p-p-f-f-f-f-f-f-f-f-f-f
```

```
                    V
p-p-f-p-p-f-f-p-f-p-f-p-f
```
The former is clearly a recently added regression, the later is a flakey failure where we happen to be testing with a commit that failed. Both will have pretty similar rates, at least according the API. I don't think that EWS should differentiate between the two, there isn't a reliable way to automatically do that.

aj062 replied:
Removed the concept of flaky when consulting results database in updated PR.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3247,8 +3251,35 @@ def commandComplete(self, cmd):
+        self._addToLog('stdio', f'\nChecking results database for failing tests, using identifier: {identifier}, configuration: {configuration}')
+        for test in failing_tests:
+            data = ResultsDatabase().is_test_flaky_or_pre_existing(test, commit=identifier, configuration=configuration)
+            self._addToLog('stdio', f"\n{test}: pass_rate: {data['pass_rate']}, pre-existing-failure={data['is_existing_failure']}, flaky={data['is_flaky_failure']}, details: {data['raw_data']}\n{data['logs']}")
```
This line ends up really long in practice, should be split.

aj062 committed:
EWS should use results database instead of doing clean-tree runs
https://bugs.webkit.org/show_bug.cgi?id=204368
rdar://82975679

Reviewed by Jonathan Bedard.

Pre-existing test failures signficantly slow down EWS, and this change attempts to speed up EWS
by consulting results database. EWS will now consult results database for the test failures which were noticed.
If those test were also failing/flaky in results db, then we ignore those failures in EWS run. Othewise, EWS will
continue with the existing logic of retry/clean-tree-run.

* Tools/CISupport/ews-build/results_db.py:
(ResultsDatabase.get_results_summary): Updated the method to not print logs, instead return them to the caller.
(ResultsDatabase.is_test_flaky_or_pre_existing): Wrapper method which contains the logic to declare if test is flaky/pre-existing or not.
* Tools/CISupport/ews-build/steps.py:
(RunWebKitTests.commandComplete):
(RunWebKitTests.filter_failures_using_results_db): Method which filters the test failures based on info from results db.
(RunWebKitTests.evaluateCommand):
(RunWebKitTests.getResultSummary):
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/257090@main
